File: test_tools.py
#!/usr/bin/env python3
"""
Test script to verify tool functionality without needing OpenAI API key
"""

import os
import tempfile
from simple_code_agent import FileTools, ShellTools, Logger

def test_file_tools():
    """Test file operations."""
    logger = Logger(verbose=True)
    file_tools = FileTools()

    logger.log("Testing File Tools", "INFO")

    # Create a temporary file for testing
    with tempfile.NamedTemporaryFile(mode='w', delete=False, suffix='.txt') as f:
        test_file = f.name

    try:
        # Test write
        logger.log("Test 1: Writing file", "INFO")
        result = file_tools.write_file(test_file, "Hello, World!\nThis is a test.")
        print(f"  Result: {result}")
        assert result['success'], "Write failed"

        # Test read
        logger.log("Test 2: Reading file", "INFO")
        result = file_tools.read_file(test_file)
        print(f"  Content: {result.get('content', 'N/A')[:50]}...")
        assert result['success'], "Read failed"
        assert "Hello, World!" in result['content'], "Content mismatch"

        # Test edit
        logger.log("Test 3: Editing file", "INFO")
        result = file_tools.edit_file(test_file, "World", "Python")
        print(f"  Result: {result}")
        assert result['success'], "Edit failed"

        # Verify edit
        result = file_tools.read_file(test_file)
        assert "Hello, Python!" in result['content'], "Edit didn't work"

        # Test list files
        logger.log("Test 4: Listing files", "INFO")
        result = file_tools.list_files(os.path.dirname(test_file) or ".")
        print(f"  Found {len(result.get('files', []))} items")
        assert result['success'], "List files failed"

        logger.log("All file tool tests passed!", "SUCCESS")

    finally:
        # Cleanup
        if os.path.exists(test_file):
            os.unlink(test_file)


def test_shell_tools():
    """Test shell command execution."""
    logger = Logger(verbose=True)
    shell_tools = ShellTools()

    logger.log("Testing Shell Tools", "INFO")

    # Test simple command
    logger.log("Test 1: Running 'echo' command", "INFO")
    result = shell_tools.run_command("echo 'Hello from shell'")
    print(f"  Output: {result.get('stdout', '').strip()}")
    assert result['success'], "Echo command failed"
    assert "Hello from shell" in result['stdout'], "Output mismatch"

    # Test command with output
    logger.log("Test 2: Running 'ls' command", "INFO")
    result = shell_tools.run_command("ls -la")
    print(f"  Lines of output: {len(result.get('stdout', '').split(chr(10)))}")
    assert result['success'], "ls command failed"

    # Test Python command
    logger.log("Test 3: Running Python command", "INFO")
    result = shell_tools.run_command("python3 -c 'print(2 + 2)'")
    print(f"  Output: {result.get('stdout', '').strip()}")
    assert result['success'], "Python command failed"
    assert "4" in result['stdout'], "Math didn't work"

    logger.log("All shell tool tests passed!", "SUCCESS")


def main():
    """Run all tests."""
    print("\n" + "="*60)
    print("Testing Simple Code Agent Tools")
    print("="*60 + "\n")

    try:
        test_file_tools()
        print()
        test_shell_tools()

        print("\n" + "="*60)
        print("✓ All tests passed successfully!")
        print("="*60 + "\n")

    except AssertionError as e:
        print(f"\n✗ Test failed: {e}\n")
        return 1
    except Exception as e:
        print(f"\n✗ Unexpected error: {e}\n")
        return 1

    return 0


if __name__ == "__main__":
    exit(main())

File: test_tools.cpp
#include "code_agent.hpp"
#include <iostream>

using namespace CodeAgent;

int main() {
    std::cout << "Testing FileTools and ShellTools..." << std::endl;

    // Test 1: Read file
    std::cout << "\n=== Test 1: Read file ===" << std::endl;
    auto result1 = FileTools::readFile("test_file.txt");
    std::cout << "Success: " << (result1.bools["success"] ? "true" : "false") << std::endl;
    std::cout << "Message: " << result1.strings["message"] << std::endl;
    if (result1.bools["success"]) {
        std::cout << "Content preview: " << result1.strings["content"].substr(0, 50) << "..." << std::endl;
    }

    // Test 2: Write file
    std::cout << "\n=== Test 2: Write file ===" << std::endl;
    auto result2 = FileTools::writeFile("test_output.txt", "This is a test output file.\n");
    std::cout << "Success: " << (result2.bools["success"] ? "true" : "false") << std::endl;
    std::cout << "Message: " << result2.strings["message"] << std::endl;

    // Test 3: List files
    std::cout << "\n=== Test 3: List files ===" << std::endl;
    auto result3 = FileTools::listFiles(".");
    std::cout << "Success: " << (result3.bools["success"] ? "true" : "false") << std::endl;
    std::cout << "Message: " << result3.strings["message"] << std::endl;
    if (result3.bools["success"]) {
        std::cout << "Files found: " << result3.arrays["files"].size() << std::endl;
        std::cout << "First 5 files:" << std::endl;
        for (size_t i = 0; i < std::min(size_t(5), result3.arrays["files"].size()); i++) {
            auto& file = result3.arrays["files"][i];
            std::cout << "  - " << file["name"] << " (" << file["type"] << ")" << std::endl;
        }
    }

    // Test 4: Run command
    std::cout << "\n=== Test 4: Run command ===" << std::endl;
    auto result4 = ShellTools::runCommand("echo 'Hello from shell'", 5);
    std::cout << "Success: " << (result4.bools["success"] ? "true" : "false") << std::endl;
    std::cout << "Message: " << result4.strings["message"] << std::endl;
    std::cout << "Output: " << result4.strings["stdout"] << std::endl;

    // Test 5: Edit file
    std::cout << "\n=== Test 5: Edit file ===" << std::endl;
    auto result5 = FileTools::editFile("test_output.txt", "test output", "MODIFIED OUTPUT");
    std::cout << "Success: " << (result5.bools["success"] ? "true" : "false") << std::endl;
    std::cout << "Message: " << result5.strings["message"] << std::endl;

    // Verify edit
    std::cout << "\n=== Verify edit ===" << std::endl;
    auto result6 = FileTools::readFile("test_output.txt");
    if (result6.bools["success"]) {
        std::cout << "Modified content: " << result6.strings["content"] << std::endl;
    }

    std::cout << "\n✓ All tool tests completed!" << std::endl;

    return 0;
}
